docs/proposals/ldap-group-sync.md

```diff
@@ -0,0 +1,108 @@
+
+```go
+// LDAPGroupDataExtractor lists the LDAP entries that correspond to the members of the given Group
+type LDAPGroupDataExtractor interface {
+	ExtractMembers(groupUID string) (members []*ldap.Entry, err error)
+	ExtractLabels(groupUID string) (labels labels.Set, err error)
+	ExtractAnnotations(groupUID string) (annotations map[string]string, err error)
+}
+```
+
+### Mapping LDAP Members to OpenShift `Identities`
+This will be done in an identical way to the mapping for [LDAP password processing](https://github.com/openshift/origin/blob/master/pkg/auth/authenticator/password/ldappassword/ldap.go#L227-243) once the requisite data is collected.
+
+```go
+// LDAPIdentityFactory creates Identites for LDAP user entries.
+type LDAPUserIdentityFactory interface {
+	IdentityFor(user *ldap.Entry) (identity userapi.Identity, err error)
+}
+```
+
+### Mapping OpenShift `Identities` Representing LDAP Members to OpenShift `Users`
+A new `UserIdentityMapper` will need to be created: the `DeterministicUserIdentityMapper`, which will assume that the unique identifier retrieved from the LDAP response and integrated into the OpenShift `Identity` will be deterministicically mapped to an OpenShift `User`. This is required as otherwise there is no deterministic mapping from the `Identity` created and any OpenShift `User`.
+
+```go
+// DeterministicUserIdentityMapper is a UserIdentityMapper that forces a deterministic mapping from
+// an Identity to a User.
+type DeterministicUserIdentityMapper struct {
+  // AllowLDAPIdentityCollisions determines if this UserIdentityMapper is allowed to make a mapping
+  // between an Identity provided by LDAP and an OpenShift User that already is mapped to another
+  // Identity provided by LDAP.
+  AllowLDAPIdentityCollisions bool
+}
+
+func (m *DeterministicUserIdentityMapper) UserFor(identity authapi.UserIdentityInfo) (user user.Info, err error) {
+}
+```
+
+## Internalizing LDAP Responses: Building OpenShift `Groups`
+### Building OpenShift `Groups` to Hold OpenShift `Users` Created or Referenced by LDAP Responses
+Given the identifying filter uesd to find a LDAP group, the corresponding OpenShift `Group` can either be found using the `externalRecordProvider` `Label` and `LDAPFilterGenerator` or created it it does not yet exist. If the group returned from the LDAP server is a first-class entry, information can be migrated to the OpenShift `Group`, but if the group is encode as attributes of first-class LDAP user entries, no information other than the name of the group can be carried over. In either case, data from the old OpenShift `Group`'s `ObjectMeta` is kept unless overridden by the update and the `Group`'s `Users` are always overriden.
```

Comment:
An LDAPGroupMapper takes an LDAP groupUID and returns the openshift group name.  This can be done as:
 1. a fixed mapping specified by the admin in a file
 1. exact groupUID match
 1. For schema1 or schema3,  an attribute match